File: inventario-service/inventario_service/src/main/kotlin/monitoramento/inventario/service/ProdutoService.java
package monitoramento.inventario.service;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

import monitoramento.inventario.repository.ProdutoRepository;

@Service
public class ProdutoService {

    private final ProdutoRepository produtoRepository;

    public ProdutoService(ProdutoRepository produtoRepository) {
        this.produtoRepository = produtoRepository;
    }

    public List<ProdutoDto> listarTodos() {
        return produtoRepository.findAll().stream().map(this::toDto).collect(Collectors.toList());
    }

    public ProdutoDto criarProduto(ProdutoDto produtoDto) {
        Produto produto = fromDto(produtoDto);
        produtoRepository.save(produto);
        return toDto(produto);
    }

    public ProdutoDto atualizarProduto(Long id, ProdutoDto produtoDto) {
        Produto produto = produtoRepository.findById(id).orElseThrow(() -> new RuntimeException("Produto não encontrado"));
        produto.setNome(produtoDto.getNome());
        produto.setPreco(produtoDto.getPreco());
        produto.setQuantidadeEmEstoque(produtoDto.getQuantidadeEmEstoque());
        produtoRepository.save(produto);
        return toDto(produto);
    }

    public void deletarProduto(Long id) {
        produtoRepository.deleteById(id);
    }

    private ProdutoDto toDto(Produto produto) {
        ProdutoDto dto = new ProdutoDto();
        dto.setId(produto.getId());
        dto.setNome(produto.getNome());
        dto.setSku(produto.getSku());
        dto.setPreco(produto.getPreco());
        dto.setQuantidadeEmEstoque(produto.getQuantidadeEmEstoque());
        dto.setCategoriaNome(produto.getCategoria().getNome());
        return dto;
    }

    private Produto fromDto(ProdutoDto dto) {
        Produto produto = new Produto();
        produto.setNome(dto.getNome());
        produto.setSku(dto.getSku());
        produto.setPreco(dto.getPreco());
        produto.setQuantidadeEmEstoque(dto.getQuantidadeEmEstoque());
        return produto;
    }
}

File: inventario-service/inventario_service/src/main/kotlin/monitoramento/inventario/service/ProdutoService.kt
package monitoramento.inventario.service;

import Produto
import ProdutoDto
import monitoramento.inventario.repository.CategoriaRepository
import monitoramento.inventario.repository.ProdutoRepository
import org.springframework.stereotype.Service

@Service
class ProdutoService(
    private val produtoRepository: ProdutoRepository,
    private val categoriaRepository: CategoriaRepository
) {

    fun listarTodos(): List<ProdutoDto> =
            produtoRepository.findAll().map { toDto(it) }

    fun criarProduto(produtoDto: ProdutoDto): ProdutoDto {
        val produto = fromDto(produtoDto)
        produtoRepository.save(produto)
        return toDto(produto)
    }

    fun atualizarProduto(id: Long, produtoDto: ProdutoDto): ProdutoDto {
        val produto = produtoRepository.findById(id).orElseThrow { RuntimeException("Produto não encontrado") }

        val categoria = categoriaRepository.findByNome(produtoDto.categoriaNome)
            ?: throw RuntimeException("Categoria não encontrada")
        val produtoAtualizado = produto.copy(
            nome = produtoDto.nome,
            sku = produtoDto.sku,
            preco = produtoDto.preco,
            quantidadeEmEstoque = produtoDto.quantidadeEmEstoque,
            //categoria = categoria
        )

        produtoRepository.save(produtoAtualizado)
        return toDto(produtoAtualizado)
    }

    fun deletarProduto(id: Long) = produtoRepository.deleteById(id)

    private fun toDto(produto: Produto): ProdutoDto {
        return ProdutoDto(
            id = produto.id,
            nome = produto.nome,
            sku = produto.sku,
            preco = produto.preco,
            quantidadeEmEstoque = produto.quantidadeEmEstoque,
            quantidadeMinimaEstoque = produto.quantidadeMinimaEstoque,
            categoriaNome = produto.categoria.nome
        )
    }

    private fun fromDto(produtoDto: ProdutoDto): Produto {
        val categoria = categoriaRepository.findByNome(produtoDto.categoriaNome)
            ?: throw RuntimeException("Categoria não encontrada")
        return Produto(
            id = produtoDto.id ?: 0L,
            nome = produtoDto.nome,
            sku = produtoDto.sku,
            preco = produtoDto.preco,
            quantidadeEmEstoque = produtoDto.quantidadeEmEstoque,
            quantidadeMinimaEstoque = produtoDto.quantidadeMinimaEstoque,
            categoria = categoria
        )
    }
}